compiler/x/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3216,6 +3215,252 @@ TR::Register *OMR::X86::TreeEvaluator::arraycopyEvaluator(TR::Node *node, TR::Co
+      }
+   // There are two cases.
+   // In the first case we know that a simple memmove or memcpy operation can
+   // be done. For this case there are 3 children:
+   //    1) The byte source pointer
+   //    2) The byte destination pointer
+   //    3) The byte count
+   //
+   // In the second case we must generate run-time tests to see if a simple
+   // byte copy can be done or if an element-by-element copy is needed.
+   // For this case there are 5 children:
+   //    1) The original source object reference
+   //    2) The original destination object reference
+   //    3) The byte source pointer
+   //    4) The byte destination pointer
+   //    5) The byte count
+   //
+
+   // ALL nodes need be evaluated to comply argument evaluation order;
+   // since size node is the last node, its evaluation can be delayed for further optimization
+   TR::Node* sizeNode = node->getLastChild();
+   for (int i = 0; i < node->getNumChildren()-1; i++)
+      {
+      cg->evaluate(node->getChild(i));
+      }
+
+   TR::Register* srcReg = cg->allocateRegister();
+   TR::Register* dstReg = cg->allocateRegister();
+   generateRegRegInstruction(MOVRegReg(), node, srcReg, node->getChild(node->getNumChildren() - 3)->getRegister(), cg);
+   generateRegRegInstruction(MOVRegReg(), node, dstReg, node->getChild(node->getNumChildren() - 2)->getRegister(), cg);
+
+   TR::DataType dt = node->getArrayCopyElementType();
+   uint32_t elementSize = 1;
+   static bool forceByteArrayElementCopy = feGetEnv("TR_ForceByteArrayElementCopy");
+   if (!forceByteArrayElementCopy)
+      {
+      if (node->isReferenceArrayCopy() || dt == TR::Address)
+         elementSize = TR::Compiler->om.sizeofReferenceField();
+      else
+         elementSize = TR::Symbol::convertTypeToSize(dt);
+      }
+
+   bool isShortConstantArray = false;
+   if (isShortConstantArray)
+      {
+      }
+   else
+      {
+      TR::Register* sizeReg = TR::TreeEvaluator::intOrLongClobberEvaluate(sizeNode, TR::TreeEvaluator::getNodeIs64Bit(sizeNode, cg), cg);
+      if (TR::Compiler->target.is64Bit() && !TR::TreeEvaluator::getNodeIs64Bit(sizeNode, cg))
+         {
+         generateRegRegInstruction(MOVZXReg8Reg4, node, sizeReg, sizeReg, cg);
+         }
+      if (elementSize == 8 && TR::Compiler->target.is32Bit())
+         {
+         arrayCopy64BitPrimitiveOnIA32(node, dstReg, srcReg, sizeReg, cg);
+         }
+      else
+         {
+         arrayCopyDefault(node, elementSize, dstReg, srcReg, sizeReg, cg);
+         }
+      cg->stopUsingRegister(sizeReg);
+      cg->decReferenceCount(sizeNode);
+      }
+
+   cg->stopUsingRegister(dstReg);
+   cg->stopUsingRegister(srcReg);
+#ifdef J9_PROJECT_SPECIFIC
+   if (node->isReferenceArrayCopy())
+      {
+      TR::TreeEvaluator::generateWrtbarForArrayCopy(node, cg);
+      }
+#endif
+   for (int i = 0; i < node->getNumChildren()-1; i++)
+      {
+      cg->decReferenceCount(node->getChild(i));
+      }
```

Comment:
Thank you for pointing this out, it will be taken care of once optimizations related to constant length array are introduced.